tests/jobserver/test_models.py

```diff
@@ -196,6 +196,335 @@ def test_job_status_unstarted_is_pending():
+def test_jobrequest_num_completed_success():
+    request = JobRequestFactory()
+
+    job1 = JobFactory(request=request, completed_at=timezone.now())
+    job2 = JobFactory(request=request, completed_at=timezone.now())
+
+    job1.needed_by = job2
+    job1.save()
+
+    assert request.num_completed == 2
+
+
+@pytest.mark.django_db
+def test_jobrequest_ordered_jobs_no_jobs():
+    assert JobRequestFactory().ordered_jobs == []
+
+
+@pytest.mark.django_db
+def test_jobrequest_ordered_jobs_one_job():
+    request = JobRequestFactory()
+
+    job = JobFactory(request=request)
+
+    assert request.ordered_jobs == [job]
+
+
+@pytest.mark.django_db
+def test_jobrequest_ordered_jobs_success():
+    request = JobRequestFactory()
+
+    job1 = JobFactory(request=request, completed_at=timezone.now())
+    job2 = JobFactory(request=request, completed_at=timezone.now(), status_code=3)
+    job3 = JobFactory(request=request, completed_at=timezone.now())
+    job4 = JobFactory(request=request, completed_at=timezone.now())
+
+    # set up hierarchy
+    job1.needed_by = job2
+    job1.save()
+
+    job2.needed_by = job4
+    job2.save()
+
+    job3.needed_by = job4
+    job3.save()
+
+    assert len(request.ordered_jobs) == 4
+    assert request.ordered_jobs == [job3, job1, job2, job4]
+
+
+@pytest.mark.django_db
+def test_jobrequest_runtime_no_jobs():
+    assert not JobRequestFactory().runtime
+
+
+@pytest.mark.django_db
+def test_jobrequest_runtime_not_completed():
+    request = JobRequestFactory()
+
+    job1 = JobFactory(
+        request=request,
+        started_at=timezone.now() - timedelta(minutes=1),
+        completed_at=timezone.now(),
+    )
+    job2 = JobFactory(request=request)
+
+    job1.needed_by = job2
+    job1.save()
+
+    assert request.started_at
+    assert not request.runtime
+
+
+@pytest.mark.django_db
+def test_jobrequest_runtime_not_started():
+    request = JobRequestFactory()
+
+    JobFactory(request=request)
+    JobFactory(request=request)
+
+    assert not request.runtime
+
+
+@pytest.mark.django_db
+def test_jobrequest_runtime_success():
+    request = JobRequestFactory()
+
+    start = timezone.now() - timedelta(hours=1)
+
+    job1 = JobFactory(
+        request=request,
+        started_at=start,
+        completed_at=start + timedelta(minutes=1),
+    )
+    job2 = JobFactory(
+        request=request,
+        started_at=start + timedelta(minutes=20),
+        completed_at=start + timedelta(minutes=21),
+    )
+    job3 = JobFactory(
+        request=request,
+        started_at=start + timedelta(minutes=40),
+        completed_at=start + timedelta(minutes=41),
+    )
+    job4 = JobFactory(
+        request=request,
+        started_at=start + timedelta(minutes=50),
+        completed_at=timezone.now(),
+    )
+
+    # set up hierarchy
+    job1.needed_by = job2
+    job1.save()
+
+    job2.needed_by = job4
+    job2.save()
+
+    job3.needed_by = job4
+    job3.save()
+
+    assert request.runtime
```

Comment:
Seems a pity not to check the runtime value in some sense here!

Reply:
I did consider it!  `Runtime` has it's own tests which check construction and output are as expected so testing that here felt like a duplication of effort.

Reply:
Then all the setting of started_at and completed_at seem redundant? Might as well lose a few lines...

Reply:
I could definitely cut it down a bit, but we're only generating a `Runtime` once a `JobRequest` has both a started_at and completed_at.